File: 6010/archive/assignment4/hw4/particle.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

// Define a structure to represent a particle
typedef struct {
    double x, y;      // Positions
    double u, v;      // Velocities
    double last_updated; // Last update time
    int cwall, cpart; // Count of collisions with walls and other particles
} Particle;

// Define a structure to represent a potential collision
typedef struct {
    int particle1, particle2; // Indices of the particles (-1 for walls)
    double time;              // Collision time
} Collision;

// Function to calculate the time until collision with a wall
double wall_collision_time(Particle *p, double L_x, double L_y, double R) {
    double time_x = (p->u > 0) ? (L_x - R - p->x) / p->u : (R - p->x) / p->u;
    double time_y = (p->v > 0) ? (L_y - R - p->y) / p->v : (R - p->y) / p->v;
    if (p->u < 0) time_x = -time_x;
    if (p->v < 0) time_y = -time_y;
    return (time_x < time_y) ? time_x : time_y; // Return the smaller time
}

// Function to calculate the time until collision between two particles
double particle_collision_time(Particle *p1, Particle *p2, double R) {
    double dx = p1->x - p2->x;
    double dy = p1->y - p2->y;
    double dvx = p1->u - p2->u;
    double dvy = p1->v - p2->v;
    
    double a = dvx*dvx + dvy*dvy;
    double b = 2 * (dx*dvx + dy*dvy);
    double c = dx*dx + dy*dy - 4*R*R;
    
    double disc = b*b - 4*a*c;
    if (disc < 0) return INFINITY; // No collision if discriminant is negative
    
    double t1 = (-b - sqrt(disc)) / (2*a);
    double t2 = (-b + sqrt(disc)) / (2*a);
    
    return (t1 < t2) ? t1 : t2; // Return the smaller time
}

// Function to update the position of a particle
void update_position(Particle *p, double time) {
    if (time > p->last_updated) {
        double dt = time - p->last_updated;
        p->x += p->u * dt;
        p->y += p->v * dt;
        p->last_updated = time;
    }
}

// Function to process a wall collision
void process_wall_collision(Particle *p, double L_x, double L_y, double R) {
    // Check which wall the collision is with and update velocity accordingly
    if (p->x <= R || p->x >= L_x - R) {
        p->u = -p->u;
        p->cwall++;
    }
    if (p->y <= R || p->y >= L_y - R) {
        p->v = -p->v;
        p->cwall++;
    }
}

// Function to process a particle collision
void process_particle_collision(Particle *p1, Particle *p2) {
    // Exchange velocities
    double temp_u = p1->u;
    double temp_v = p1->v;
    p1->u = p2->u;
    p1->v = p2->v;
    p2->u = temp_u;
    p2->v = temp_v;
    
    // Update collision counts
    p1->cpart++;
    p2->cpart++;
}

// Function to read particles from a file
Particle *read_particles(const char *filename, int *N, double *R, double *L_x, double *L_y) {
    FILE *file = fopen(filename, "r");
    if (!file) {
        fprintf(stderr, "Error opening file.\n");
        exit(EXIT_FAILURE);
    }
    
    fscanf(file, "%d", N); // Number of particles
    fscanf(file, "%lf", R); // Radius
    fscanf(file, "%lf %lf", L_x, L_y); // Box dimensions
    
    Particle *particles = malloc(*N * sizeof(Particle));
    if (!particles) {
        fprintf(stderr, "Error allocating memory.\n");
        fclose(file);
        exit(EXIT_FAILURE);
    }
    
    for (int i = 0; i < *N; ++i) {
        Particle *p = &particles[i];
        fscanf(file, "%lf %lf %lf %lf", &p->x, &p->y, &p->u, &p->v);
        p->last_updated = 0;
        p->cwall = p->cpart = 0;
    }
    
    fclose(file);
    return particles;
}

// Function to sort collisions by time
void sort_collisions(Collision *collisions, int count) {
    // Simple insertion sort
    for (int i = 1; i < count; ++i) {
        Collision temp = collisions[i];
        int j = i - 1;
        while (j >= 0 && collisions[j].time > temp.time) {
            collisions[j + 1] = collisions[j];
            j--;
        }
        collisions[j + 1] = temp;
    }
}

// Main simulation function
void simulate(Particle *particles, int N, double L_x, double L_y, double R, double end_time) {
    // Calculate the number of potential collisions
    int collision_count = N + (N*(N-1))/2;
    Collision *collisions = malloc(collision_count * sizeof(Collision));
    if (!collisions) {
        fprintf(stderr, "Error allocating memory for collisions.\n");
        free(particles);
        exit(EXIT_FAILURE);
    }
    
    // Initial calculation of potential collision times
    for (int i = 0; i < N; ++i) {
        // Calculate wall collision time
        collisions[i].particle1 = i;
        collisions[i].particle2 = -1; // Indicator for wall collision
        collisions[i].time = particles[i].last_updated + wall_collision_time(&particles[i], L_x, L_y, R);
        
        // Calculate particle-particle collision times
        for (int j = i + 1; j < N; ++j) {
            int index = N + i * (N - 1) / 2 + j - (i + 1);
            collisions[index].particle1 = i;
            collisions[index].particle2 = j;
            collisions[index].time = particles[i].last_updated + particle_collision_time(&particles[i], &particles[j], R);
        }
    }
    
    // Sort potential collisions by time
    sort_collisions(collisions, collision_count);
    
    double current_time = 0.0;
    while (current_time < end_time && collision_count > 0) {
        // Process the earliest potential collision
        Collision collision = collisions[0];
        current_time = collision.time;
        if (current_time > end_time) break; // Stop if we've reached the end time
        
        // Update particle positions to the collision time
        for (int i = 0; i < N; ++i) {
            update_position(&particles[i], current_time);
        }
        
        if (collision.particle2 == -1) { // Wall collision
            process_wall_collision(&particles[collision.particle1], L_x, L_y, R);
        } else { // Particle collision
            process_particle_collision(&particles[collision.particle1], &particles[collision.particle2]);
        }
        
        // Re-calculate potential collisions involving the colliding particles
        for (int i = 0; i < N; ++i) {
            if (i == collision.particle1 || i == collision.particle2) {
                collisions[i].time = current_time + wall_collision_time(&particles[i], L_x, L_y, R);
                for (int j = 0; j < N; ++j) {
                    if (j != i) {
                        int index = N + i * (N - 1) / 2 + j - (i + 1);
                        collisions[index].time = current_time + particle_collision_time(&particles[i], &particles[j], R);
                    }
                }
            }
        }
        
        // Sort the array of potential collisions again
        sort_collisions(collisions, collision_count);
    }
    
    // Advance all particle positions to the end time and output their final states
    for (int i = 0; i < N; ++i) {
        update_position(&particles[i], end_time);
        printf("%.6f, %.6f, %d, %d\n", particles[i].x, particles[i].y, particles[i].cwall, particles[i].cpart);
    }
    
    // Free allocated memory
    free(collisions);
    free(particles);
}

int main(int argc, char *argv[]) {
    if (argc != 3) {
        fprintf(stderr, "Usage: %s <filename> <end_time>\n", argv[0]);
        return EXIT_FAILURE;
    }

    const char *filename = argv[1];
    double end_time = atof(argv[2]);

    if (end_time <= 0) {
        fprintf(stderr, "End time must be positive.\n");
        return EXIT_FAILURE;
    }

    int N; // Number of particles
    double R, L_x, L_y; // Radius and box dimensions

    // Read particles from the file
    Particle *particles = read_particles(filename, &N, &R, &L_x, &L_y);

    // Perform the simulation
    simulate(particles, N, L_x, L_y, R, end_time);

    return EXIT_SUCCESS;
}

File: 6010/archive/assignment4/hw4/particle.py
import math
import itertools
import sys

# Class to represent a particle
class Particle:
    def __init__(self, x, y, vx, vy, radius):
        self.x = x
        self.y = y
        self.vx = vx
        self.vy = vy
        self.radius = radius
        self.last_update_time = 0.0
        self.wall_collision_count = 0
        self.particle_collision_count = 0

    def update_position(self, time):
        elapsed_time = time - self.last_update_time
        self.x += self.vx * elapsed_time
        self.y += self.vy * elapsed_time
        self.last_update_time = time

    def collide_with_wall(self, wall):
        if wall == 'vertical':
            self.vx = -self.vx
        else:
            self.vy = -self.vy
        self.wall_collision_count += 1

    def collide_with_particle(self, other):
        self.vx, other.vx = other.vx, self.vx
        self.vy, other.vy = other.vy, self.vy
        self.particle_collision_count += 1
        other.particle_collision_count += 1

# Class to represent a collision
class Collision:
    def __init__(self, time, particles, collision_type):
        self.time = time
        self.particles = particles
        self.collision_type = collision_type

# Class to simulate the particle dynamics
class Simulation:
    def __init__(self, particles, box_length_x, box_length_y):
        self.particles = particles
        self.box_length_x = box_length_x
        self.box_length_y = box_length_y
        self.collisions = []

    def calculate_wall_collisions(self, particle):
        times_to_collision = {'left': float('inf'), 'right': float('inf'),
                              'bottom': float('inf'), 'top': float('inf')}
        if particle.vx > 0:
            times_to_collision['right'] = (self.box_length_x - particle.radius - particle.x) / particle.vx
        elif particle.vx < 0:
            times_to_collision['left'] = (particle.radius - particle.x) / particle.vx
        if particle.vy > 0:
            times_to_collision['top'] = (self.box_length_y - particle.radius - particle.y) / particle.vy
        elif particle.vy < 0:
            times_to_collision['bottom'] = (particle.radius - particle.y) / particle.vy

        earliest_collision_time = min(times_to_collision.values())
        collision_wall = [wall for wall, time in times_to_collision.items() if time == earliest_collision_time][0]
        return Collision(earliest_collision_time + particle.last_update_time, [particle], collision_wall)

    def calculate_particle_collisions(self, particle, other):
        dx = other.x - particle.x
        dy = other.y - particle.y
        dvx = other.vx - particle.vx
        dvy = other.vy - particle.vy

        a = dvx**2 + dvy**2
        b = 2 * (dx * dvx + dy * dvy)
        c = dx**2 + dy**2 - (2 * particle.radius)**2

        discriminant = b**2 - 4 * a * c
        if discriminant < 0:
            return Collision(float('inf'), [particle, other], 'particle')

        discriminant_sqrt = math.sqrt(discriminant)
        t1 = (-b - discriminant_sqrt) / (2 * a)
        t2 = (-b + discriminant_sqrt) / (2 * a)

        if t1 > 0:
            collision_time = t1
        elif t2 > 0:
            collision_time = t2
        else:  # both times are in the past
            return Collision(float('inf'), [particle, other], 'particle')

        return Collision(collision_time + particle.last_update_time, [particle, other], 'particle')

    def simulate(self, end_time):
        # Initialize potential collisions
        for particle in self.particles:
            self.collisions.append(self.calculate_wall_collisions(particle))
        for particle, other in itertools.combinations(self.particles, 2):
            self.collisions.append(self.calculate_particle_collisions(particle, other))

        # Process collisions until end time
        while True:
            self.collisions.sort(key=lambda collision: collision.time)
            next_collision = self.collisions.pop(0)
            if next_collision.time > end_time:
                break
            for particle in next_collision.particles:
                particle.update_position(next_collision.time)
            if next_collision.collision_type == 'vertical' or next_collision.collision_type == 'horizontal':
                next_collision.particles[0].collide_with_wall(next_collision.collision_type)
            else:
                next_collision.particles[0].collide_with_particle(next_collision.particles[1])

            # Recalculate potential collisions involving the colliding particles
            self.collisions = [collision for collision in self.collisions if not any(p in collision.particles for p in next_collision.particles)]
            for particle in next_collision.particles:
                self.collisions.append(self.calculate_wall_collisions(particle))
                for other in [p for p in self.particles if p != particle]:
                    self.collisions.append(self.calculate_particle_collisions(particle, other))

        # Update all particle positions to the end time
        for particle in self.particles:
            particle.update_position(end_time)

    def get_simulation_results(self):
        return [(particle.x, particle.y, particle.wall_collision_count, particle.particle_collision_count) for particle in self.particles]

if __name__ == "__main__":
    if len(sys.argv) != 3:
        print("Usage: python3 particle.py <filename> <end_time>")
        sys.exit(1)

    filename = sys.argv[1]
    end_time = float(sys.argv[2])
    
    # Read the file to get particle data
    with open(filename, 'r') as file:
        N = int(file.readline().strip())
        radius = float(file.readline().strip())
        box_length_x, box_length_y = map(float, file.readline().strip().split())
        particles = []
        for line in file:
            x, y, vx, vy = map(float, line.strip().split())
            particles.append(Particle(x, y, vx, vy, radius))

    # Run the simulation
    simulation = Simulation(particles, box_length_x, box_length_y)
    simulation.simulate(end_time)
    results = simulation.get_simulation_results()

    # Print the results
    for result in results:
        print("{:.6f}, {:.6f}, {}, {}".format(*result))
        
# Example usage:
# particles = [Particle(x, y, vx, vy, radius) for x, y, vx, vy in initial_conditions]
# simulation = Simulation(particles, box_length_x, box_length_y)
# simulation.simulate(end_time)
# results = simulation.get_simulation_results()
